docs/Contributing/Fleet-UI-Testing.md

```diff
@@ -0,0 +1,362 @@
+
+describe("Radio - component", () => {
+  it("renders the radio label text from the label prop", () => {
+    render(
+      <Radio
+        checked
+        label={"Radio Label"}
+        value={"radioValue"}
+        id={"test-radio"}
+        onChange={noop}
+      />
+    );
+
+    const labelText = screen.getByText("Radio Label");
+    expect(labelText).toBeInTheDocument();
+  });
+
+  it("passes the radio input value when checked", async () => {
+    const user = userEvent.setup();
+    const changeHandlerSpy = jest.fn();
+
+    render(
+      <Radio
+        label={"Radio Label"}
+        value={"radioValue"}
+        id={"test-radio"}
+        onChange={changeHandlerSpy}
+      />
+    );
+
+    const radio = screen.getByRole("radio", { name: "Radio Label" });
+    await user.click(radio);
+
+    expect(changeHandlerSpy).toHaveBeenCalled();
+    expect(changeHandlerSpy).toHaveBeenCalledWith("radioValue");
+  });
+});
+```
+
+##### Reusable Hooks Testing
+
+// TODO
+
+#### Integration Testing
+
+We use integration testing to strike a balance between speed and expense to write our tests. We use
+them to test multiple reusable components that come together into less reusable App Widgets. We also try to use
+minimal mocking, but can mock at the backend level if required.
+
+##### App widget Testing
+
+This layer of tests is great for testing difficult to obtain states and edge cases as the test setup
+is simpler than e2e tests. We highly utilise react-testing-library to interface with these components.
+
+[source here](https://github.com/fleetdm/fleet/blob/main/frontend/components/forms/ResetPasswordForm/ResetPasswordForm.tests.jsx)
+
+```tsx
+import React from "react";
+import { render, screen } from "@testing-library/react";
+
+import { renderWithSetup } from "test/testingUtils";
+
+import ResetPasswordForm from "./ResetPasswordForm";
+
+describe("ResetPasswordForm - component", () => {
+  const newPassword = "password123!";
+  const submitSpy = jest.fn();
+
+  it("does not submit the form if the password is invalid", async () => {
+    const invalidPassword = "invalid";
+    const { user } = renderWithSetup(
+      <ResetPasswordForm handleSubmit={submitSpy} />
+    );
+
+    await user.type(screen.getByLabelText("New password"), invalidPassword);
+    await user.type(screen.getByLabelText("Confirm password"), invalidPassword);
+    await user.click(screen.getByRole("button", { name: "Reset password" }));
+
+    const passwordError = screen.getByText(
+      "Password must meet the criteria below"
+    );
+    expect(passwordError).toBeInTheDocument();
+    expect(submitSpy).not.toHaveBeenCalled();
+  });
+});
+```
+
+#### E2E testing
+
+Our e2e layer tests all the systems of the software (frontend and backend) together to
+ensure the application works as intended. To support this we do not often mock API responses at this
+layer of testing (with an exception for mocking network error responses). We want to test the
```

Comment:
Remove parentheses and create a separate sentence so we can include the external APIs exception:

"Exceptions include mocking network error responses and mocking external APIs such as Jira or Zendesk integration responses."